src/openrct2/world/Sprite.cpp

```diff
@@ -613,6 +614,60 @@ void sprite_misc_update_all()
     }
 }
 
+// Performs a search to ensure that insert keeps next_in_quadrant in sprite_index order
+static void SpriteSpatialInsert(SpriteBase* sprite, CoordsXY newLoc)
+{
+    size_t newIndex = GetSpatialIndexOffset(newLoc.x, newLoc.y);
+
+    auto* next = &gSpriteSpatialIndex[newIndex];
+    while (sprite->sprite_index < *next && *next != SPRITE_INDEX_NULL)
+    {
+        auto sprite2 = &get_sprite(*next)->generic;
+        next = &sprite2->next_in_quadrant;
+    }
+
+    sprite->next_in_quadrant = *next;
+    *next = sprite->sprite_index;
+}
+
+static void SpriteSpatialRemove(SpriteBase* sprite)
+{
+    size_t currentIndex = GetSpatialIndexOffset(sprite->x, sprite->y);
+    auto* index = &gSpriteSpatialIndex[currentIndex];
+
+    // This indicates that the spatial index data is all incorrect.
+    // Suggest rebuilding the sprite spatial index.
+    Guard::Assert(*index != SPRITE_INDEX_NULL, "Bad sprite spatial Index");
+
+    auto* sprite2 = &get_sprite(*index)->generic;
+    while (sprite != sprite2)
+    {
+        index = &sprite2->next_in_quadrant;
+        if (*index == SPRITE_INDEX_NULL)
+        {
+            break;
+        }
+        sprite2 = &get_sprite(*index)->generic;
+    }
+    *index = sprite->next_in_quadrant;
+}
+
+static void SpriteSpatialMove(SpriteBase* sprite, CoordsXY newLoc)
+{
+    if (!map_is_location_valid(newLoc))
+    {
+        newLoc.setNull();
```

Comment:
I feel like this will cause `GetSpatialIndexOffset` to throw an assert. Perhaps just remove it from the index and not try to add it?

Reply:
It still needs to be in the index as it needs to end up in `SPATIAL_INDEX_LOCATION_NULL`.
It should work fine performing setNull() as the value is the same as LOCATION_NULL. Although I guess thats not guaranteed to be the same.

Reply:
How about just removing this check and instead have `GetSpatialIndexOffset` return `SPATIAL_INDEX_LOCATION_NULL` if it is out of range?